devdoc/channel.md

```diff
@@ -0,0 +1,267 @@
+
+**SRS_CHANNEL_43_059: [** If the current `state` is `CHANNEL_PUSHING`, `channel_pull` shall wait for the `state` to change by calling `wait_on_address`. **]**
+
+**SRS_CHANNEL_43_012: [** If the current `state` is `CHANNEL_PUSHED`, `channel_pull` shall : **]**
+
+- **SRS_CHANNEL_43_060: [** store `push_callback`, `push_context` and `data` given to `channel_push`. **]**
+
+- **SRS_CHANNEL_43_042: [** call `interlocked_compare_exchange` to increment the `EPOCH` and change the state to `CHANNEL_READY`. **]**
+
+- **SRS_CHANNEL_43_013: [** call the given `pull_callback` with `pull_context` as the given `pull_context`, `data` as the stored `data` and `result` as `CHANNEL_RESULT_OK`. **]**
+
+- **SRS_CHANNEL_43_053: [** release the reference to the stored `data` by calling `THANDLE_ASSIGN`. **]**
+
+- **SRS_CHANNEL_43_014: [** call the stored `push_callback` with the stored `push_context` and `result` as `CHANNEL_RESULT_OK`. **]**
+
+- **SRS_CHANNEL_43_054: [** set `out_op_pull` to `NULL` by calling `THANDLE_ASSIGN`. **]**
+
+- **SRS_CHANNEL_43_015: [** return `CHANNEL_RESULT_SYNC`. **]**
+
+**SRS_CHANNEL_43_016: [** If the current state is `CHANNEL_READY`, `channel_pull` shall: **]**
+
+**SRS_CHANNEL_43_061: [** set the state to `CHANNEL_PULLING` by calling `interlocked_compare_exchange`. **]**
+
+- **SRS_CHANNEL_43_017: [** call `async_op_create` with `cancel_pull` as the cancel function. **]**
+
+- **SRS_CHANNEL_43_018: [** store the given `pull_callback` and the given `pull_context` in the `channel`. **]**
+
+- **SRS_CHANNEL_43_019: [** store `state` in the created `THANDLE(ASYNC_OP)`. **]**
+
+- **SRS_CHANNEL_43_020: [** store the given `channel` in the created `THANDLE(ASYNC_OP)` by calling `THANDLE_INITIALIZE`. **]**
+
+- **SRS_CHANNEL_43_021: [** store the created `THANDLE(ASYNC_OP)` in `out_op_pull` by calling `THANDLE_INITIALIZE_MOVE`. **]**
+
+- **SRS_CHANNEL_43_062: [** set the `state` of the given `channel` to `CHANNEL_PULLED` by calling `interlocked_exchange`. **]**
+
+- **SRS_CHANNEL_43_063: [** signal waiting threads by calling `wake_by_address_all`. **]**
+
+- **SRS_CHANNEL_43_022: [** return `CHANNEL_RESULT_ASYNC`. **]**
+
+**SRS_CHANNEL_43_023: [** If there are any failures, `channel_pull` shall fail and return `CHANNEL_RESULT_ERROR`. **]**
+
+### cancel_pull
+```
+static void cancel_pull(void* context)
+```
+
+`cancel_pull` is the cancel function given to `async_op_create` when `channel_pull` is called.
+
+**SRS_CHANNEL_43_046: [** If the current `EPOCH` and `state` of the `channel` is the same as the initial `EPOCH` and `state` of the `channel` when `channel_pull` was called, `cancel_pull` shall: **]**
+
+- **SRS_CHANNEL_43_047: [** increment the `EPOCH` and set the `state` of the `channel` stored in the given `context` to `CHANNEL_READY` by calling `interlocked_compare_exchange`. **]**
```

Comment:
>interlocked_compare_exchange [](http://example.com/codeflow?start=148&length=28)

This is the last time I am making this comment, but interlocked_`compare`_exchange is essentially an "if" + "exchange"  instruction. Not sure what the "if" is here and how it is used.

Here and everywhere, I have a feeling this comment might apply to other places in this document.

Reply:
43_046 says : If ...., `cancel_pull` shall:

This is the if condition.

Indented requirements (47-48), are the requirements that execute if the condition is true. 

For the "else" part, do nothing is implicit.

Reply:
Indenting conditional requirements for better readability